mailpoet/lib/Automation/Integrations/MailPoet/Actions/SendWelcomeEmailAction.php

```diff
@@ -0,0 +1,119 @@
+  private $welcomeScheduler;
+
+  /** @var NewslettersRepository */
+  private $newslettersRepository;
+
+  /** @var ScheduledTasksRepository */
+  private $scheduledTasksRepository;
+
+  public function __construct(
+    SubscribersRepository $subscribersRepository,
+    WelcomeScheduler $welcomeScheduler,
+    NewslettersRepository $newslettersRepository,
+    ScheduledTasksRepository $scheduledTasksRepository
+  ) {
+    $this->subscribersRepository = $subscribersRepository;
+    $this->welcomeScheduler = $welcomeScheduler;
+    $this->newslettersRepository = $newslettersRepository;
+    $this->scheduledTasksRepository = $scheduledTasksRepository;
+  }
+
+  public function getKey(): string {
+    return 'mailpoet:sendWelcomeEmail'; // TODO: casing for multi word keys? send-welcome-email? send_welcome_email?
+  }
+
+  public function getName(): string {
+    return __('Send Welcome Email', 'mailpoet');
+  }
+
+  public function run(Workflow $workflow, WorkflowRun $workflowRun, Step $step): void {
+    $subscriberSubject = $workflowRun->getSubjects()['mailpoet:subscriber'] ?? null;
+    if (!$subscriberSubject instanceof SubscriberSubject) {
+      throw InvalidStateException::create()->withMessage(__('No mailpoet:subscriber subject provided.', 'mailpoet'));
+    }
+
+    $segmentSubject = $workflowRun->getSubjects()['mailpoet:segment'] ?? null;
+    if (!$segmentSubject instanceof SegmentSubject) {
+      throw InvalidStateException::create()->withMessage(__('No mailpoet:segment subject provided.', 'mailpoet'));
+    }
+
+    $globalSubscriberStatus = $subscriberSubject->getSubscriberStatusField()->getFactory()();
+    if ($globalSubscriberStatus !== SubscriberEntity::STATUS_SUBSCRIBED) {
+      throw InvalidStateException::create()->withMessage(__(sprintf("Cannot send a welcome email to a subscriber with a global subscription status of '%s'.", $globalSubscriberStatus), 'mailpoet'));
+    }
+
+    $segmentName = $segmentSubject->getNameField()->getFactory()();
+    $idField = $subscriberSubject->getSubscriberIdField();
+    $subscriberId = $idField->getFactory()();
+    if ($subscriberId === null) {
+      throw NotFoundException::create()->withMessage(__(sprintf("Subscriber with ID '%s' not found.", $subscriberId), 'mailpoet'));
+    }
+    
+    $subscriber = $this->subscribersRepository->findOneById($subscriberId);
+    // This is for PHPStan, which doesn't understand that retrieving the ID above means the subscriber entity exists
+    if (!$subscriber instanceof SubscriberEntity) {
+      throw NotFoundException::create()->withMessage(__(sprintf("Subscriber with ID '%s' not found.", $subscriberId), 'mailpoet'));
+    }
+
+    // TODO: Are these necessarily "subscribed" status? Maybe there's a better way to check this
+    $subscriberSegments = $subscriber->getSegments();
+    $subscribedSegmentNames = array_map(function(SegmentEntity $segment) {
+      return $segment->getName();
+    }, $subscriberSegments);
```

Comment:
They are not only "subscribed" status, I think. Also see how the `mailpoet_segment_subscribed` [was implemented](https://github.com/mailpoet/mailpoet/commit/86d934ba363b39bf36d9d090a7914ef833e24078) — checks for both subscriber status and subscriber-segment status will be needed.

Reply:
I believe I now have both of these checks in place, with a separate check for the global status and then using `findSubscribersInSegments` for the segment specific checks.